X2WOTCCommunityPromotionScreen/Src/X2WOTCCommunityPromotionScreen/Classes/CPS_UIArmory_PromotionHero.uc

```diff
@@ -895,217 +817,328 @@ function bool CanPurchaseAbility(int Rank, int Branch, name AbilityName)
 	///
 	///```event
-	/// EventID: OverrideCanPurchaseAbility,
-	/// EventData: [inout bool bCanPurchaseAbility,
-	///				inout string strLocReasonLocked,
-	///				in name nReasonLocked,
-	///				in name AbilityTemplateName, 
-	///				in int Rank, in int Row, 
-	///				in int AbilitiesPerRank, 
-	///				in bool bAsResistanceHero, 
-	///				in bool bCanSpendAP],
+	/// EventID: CPS_OverrideCanPurchaseAbilityProperties,
+	/// EventData: [in name AbilityTemplateName,
+	/// 			in int iRank,
+	///				in int iRow,
+	///				in bool bPowerfulAbility,
+	///				in bool bColonelRankAbility,
+	/// 			inout bool bClassAbility,
+	///				inout bool bUnitMeetsAbilityPrerequisites,
+	///				inout bool bUnitHasPurchasedClassPerkAtRank,
+	///				inout bool bUnitMeetsRankRequirement,
+	///				inout bool bUnitCanSpendAP,
+	///				in bool bPromotionFreeUnlock,
+	///				in bool bAsResistanceHero,
+	///				in int AbilitiesPerRank],
 	/// EventSource: XComGameState_Unit (UnitState),
 	/// NewGameState: none
-	///```
+	///```	
 	Tuple = new class'XComLWTuple';
-	Tuple.Id = 'OverrideCanPurchaseAbility';
-	Tuple.Data.Add(9);
+	Tuple.Id = 'CPS_OverrideCanPurchaseAbilityProperties';
 
-	Tuple.Data[0].kind = XComLWTVBool;
-	Tuple.Data[0].b = nReasonLocked == ''; // bCanPurchaseAbility
-	Tuple.Data[1].kind = XComLWTVString;
-	Tuple.Data[1].s = strLocReasonLocked;
-	Tuple.Data[2].kind = XComLWTVName;
-	Tuple.Data[2].n = nReasonLocked;
-	Tuple.Data[3].kind = XComLWTVName;
-	Tuple.Data[3].n = AbilityName;
-	Tuple.Data[4].kind = XComLWTVInt;
-	Tuple.Data[4].i = Rank;
-	Tuple.Data[5].kind = XComLWTVInt;
-	Tuple.Data[5].i = Branch;
-	Tuple.Data[6].kind = XComLWTVInt;
-	Tuple.Data[6].i = AbilitiesPerRank;
-	Tuple.Data[7].kind = XComLWTVBool;
-	Tuple.Data[7].b = bAsResistanceHero;
-	Tuple.Data[8].kind = XComLWTVBool;
-	Tuple.Data[8].b = CanSpendAP();
+	FillTupleFromAbilityMetaInfo(MetaInfo, Tuple);
+
+	`XEVENTMGR.TriggerEvent(Tuple.Id, Tuple, UnitState);
+
+	UpdateAbilityMetaInfoFromTuple(MetaInfo, Tuple);
+}
+
+private function bool TriggerOverrideCanPurchaseAbility(CPSAbilityMetaInfo MetaInfo, XComGameState_Unit UnitState, const bool bCanPurchaseAbility, out string strLocReasonLocked)
+{
+	local XComLWTuple Tuple;
+
+	/// Mods can listen to the 'CPS_OverrideCanPurchaseAbility' event to use their own logic 
+	/// to make the final decision whether this unit should be able to unlock this ability.
+	///
+	/// - If a mod wants to allow unlocking this ability, then setting `bCanPurchaseAbility = true;` is enough.
+	/// - If a mod wants to disallow unlocking this ability, then setting `bCanPurchaseAbility = false;` is enough,
+	///	though ideally the mod should also provide a `strLocReasonLocked` string that will be displayed in the UI
+	/// as a reason why this ability cannot be unlocked at this time.
+	///
+	///```event
+	/// EventID: CPS_OverrideCanPurchaseAbility,
+	/// EventData: [in name AbilityTemplateName,
+	/// 			in int iRank,
```

Comment:
Although most folks will read this with 1 tab == 4 spaces, it's probably still worth making the indent consistently spaces or tabs for these properties, particularly if these are turned into HTML docs at some point in the future.

Reply:
I'll get that fixed. Anything else catch your eye?